homeassistant/components/azure_event_hub/__init__.py

```diff
@@ -1,89 +1,213 @@
+        }
+        conn_str_client = False
+
+    instance = hass.data[DOMAIN] = AEHThread(
+        hass, client_args, conn_str_client, entities_filter
     )
-    async_sender = client.add_async_sender()
-    await client.run_async()
-
-    encoder = JSONEncoder()
-
-    async def async_send_to_event_hub(event: Event):
+    instance.async_initialize()
+
+    async def async_unload_entry(hass: HomeAssistant):
+        """Shut down the AEH Thread."""
+        instance.queue.put(None)
+        instance.join()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_unload_entry)
+
+    instance.start()
+    return await instance.async_ready
+
+
+class AEHThread(threading.Thread):
+    """A threaded event handler class."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client_args: Dict[str, Any],
+        conn_str_client: bool,
+        entities_filter: vol.Schema,
+    ):
+        """Initialize the listener."""
+        threading.Thread.__init__(self, name="AzureEventHub")
+        self.hass = hass
+        self.queue = queue.Queue()
+        self._client_args = client_args
+        self._conn_str_client = conn_str_client
+        self._entities_filter = entities_filter
+        self.async_ready = asyncio.Future()
+        self.write_errors = 0
+        self.shutdown = False
+        # suppress the INFO and below logging on the underlying packages, they are very verbose, even at INFO
+        logging.getLogger("uamqp").setLevel(logging.WARNING)
+        logging.getLogger("azure.eventhub.client_abstract").setLevel(logging.WARNING)
+
+    def event_to_filtered_event_data(self, event: Event):
         """Send states to Event Hub."""
         state = event.data.get("new_state")
         if (
             state is None
             or state.state in (STATE_UNKNOWN, "", STATE_UNAVAILABLE)
-            or not entities_filter(state.entity_id)
+            or not self._entities_filter(state.entity_id)
         ):
             return
-
-        event_data = EventData(
-            json.dumps(obj=state.as_dict(), default=encoder.encode).encode("utf-8")
+        event_data = EventData(json.dumps(obj=state, cls=JSONEncoder).encode("utf-8"))
+        return event_data
+
+    def _event_listener(self, event: Event):
+        """Listen for new messages on the bus and queue them for AEH."""
+        item = (time.monotonic(), event)
+        self.queue.put(item)
+
+    @callback
+    def async_initialize(self):
+        """Initialize the recorder."""
+        self.hass.bus.async_listen(MATCH_ALL, self._event_listener)
+
+    @staticmethod
+    def batch_timeout():
+        """Return number of seconds to wait for more events."""
+        return BATCH_TIMEOUT
+
+    def fill_batch(self, event_data_batch: EventDataBatch):
+        """Return a batch of events formatted for writing."""
+        dequeue_count = 0
+        dropped = 0
+        can_add = True
+        try:  # pylint: disable=too-many-nested-blocks
```

Comment:
Please only wrap the line(s) that can raise. Wrap the `self.queue.get` call and `break` in the `except` block.